File: src/windows/QRScanner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Runtime.Serialization;
using System.Windows;
using Microsoft.Phone.Tasks;
using WPCordovaClassLib.Cordova;
using WPCordovaClassLib.Cordova.Commands;
using WPCordovaClassLib.Cordova.JSON;
using Microsoft.Phone.Info;
using System.IO;
using System.IO.IsolatedStorage;
using Microsoft.Devices;

    public class QRScanner : BaseCommand
    {

        enum QRScannerError
        {
            UNEXPECTED_ERROR,
            CAMERA_ACESS_DENIED,
            CAMERA_ACESS_RESTRICTED,
            BACK_CAMERA_UNAVAILABLE,
            FRONT_CAMERA_UNAVAILABLE,
            CAMERA_UNAVAILABLE,
            SCAN_CANCELED,
            LIGHT_UNAVAILABLE,
            OPEN_SETTINGS_UNAVAILABLE
        }
        /**
         * Error Callbacks:
         * DispatchCommandResult(new PluginResult(PluginResult.Status.ERROR,
         *      Convert.ToString((int)QRScannerError.BACK_CAMERA_UNAVAILABLE)));
         *
         * Success Callbacks:
         * DispatchCommandResult(); (success - no PluginResult object)
         * DispatchCommandResult(new PluginResult(PluginResult.Status.OK, "Dict"));
         *
         * */
         /**

        private void error(string error)
        {
            DispatchCommandResult(new PluginResult(PluginResult.Status.OK,
                Convert.ToString((int)QRScannerError.error)));
        }

        private void success()
        {
            DispatchCommandResult();
        }
        **/

        //<---Begin External API--->
        public void prepare(string options)
        {

        }

        public void scan(string options)
        {
            //Cordova.plugin.BarcodeScanner.scan(success, error("SCAN_CANCELED"));
        }

        public void show(string options)
        {
            DispatchCommandResult(new PluginResult(PluginResult.Status.OK, "yeehaw");
        }

        public void hide(string options)
        {

        }

        public void enableLight(string options)
        {

        }

        public void disableLight(string options)
        {

        }

        public void useFrontCamera(string options)
        {

        }

        public void useBackCamera(string options)
        {

        }

        public void pausePreview(string options)
        {

        }

        public void resumePreview(string options)
        {

        }

        public void openSettings(string options)
        {
            Windows.System.Launcher.LaunchUriAsync(new Uri("ms-settings-wifi:"));
        }

        public void getStatus(string dontUseMe)
        {
            bool authorized = false, denied = false, restricted = false,
                prepared = false, showing = false, scanning = false,
                previewing = false, lightEnabled = false, canOpenSettings =false,
                canEnableLight = false, canChangeCamera = false;

            int currentCamera = 1;

            string statusDict = String.Format("\"authorized\":\"{0}\",\"denied\":\"{1}\",\"restricted\":\"{2}\",\"prepared\":\"{3}\",\"showing\":\"{4}\",\"scanning\":\"{5}\",\"previewing\":{6}\",\"lightEnabled\":\"{7}\",\"canOpenSettings\":\"{8}\",\"canEnableLight\":\"{9}\",\"canChangeCamera\":\"{10}\",\"currentCamera\":\"{11}",
                                        Convert.ToInt32(authorized),
                                        Convert.ToInt32(denied),
                                        Convert.ToInt32(restricted),
                                        Convert.ToInt32(prepared),
                                        Convert.ToInt32(showing),
                                        Convert.ToInt32(scanning),
                                        Convert.ToInt32(previewing),
                                        Convert.ToInt32(lightEnabled),
                                        Convert.ToInt32(canOpenSettings),
                                        Convert.ToInt32(canEnableLight),
                                        Convert.ToInt32(canChangeCamera),
                                        currentCamera);

            DispatchCommandResult(new PluginResult(PluginResult.Status.OK, "{" + statusDict + "}"));
        }
    }

File: src/windows/lib/qrScanner.js
const preview = require('./preview');
const barcodeReader = new QRReader.Reader();
const VideoCapture = QRReader.VideoCapture;
const Promise = WinJS.Promise;

const errorTypes = {
  UNEXPECTED_ERROR: 0,
  CAMERA_ACESS_DENIED: 1,
  CAMERA_ACCESS_RESTRICTED: 2,
  BACK_CAMERA_UNAVAILABLE: 3,
  FRONT_CAMERA_UNAVAILABLE: 4,
  CAMERA_UNAVAILABLE: 5,
  SCAN_CANCELED: 6,
  LIGHT_UNAVAILABLE: 7,
  OPEN_SETTINGS_UNAVAILABLE: 8
};

const cameraTypes = {
  BACK: 0,
  FRONT: 1
};

const defaultStatusFlags = {
  prepared: false,
  authorized: false,
  denied: false,
  restricted: false,
  scanning: false,
  previewing: false,
  showing: false,
  lightEnabled: false,
  canOpenSettings: false,
  canEnableLight: false,
  canChangeCamera: false,
  currentCamera: cameraTypes.BACK
};

let statusFlags;
let currentVideoCapture;

function resetStatusFlags() {

  statusFlags = {};
  for (let property in defaultStatusFlags) {
    statusFlags[property] = defaultStatusFlags[property];
  }
  return statusFlags;
}

function reset() {
  document.body.removeEventListener('click', onPreviewClick);
  preview.destroy();
  if (currentVideoCapture) currentVideoCapture.destroy();
  currentVideoCapture = null;
  availableCameras = null;
  resetStatusFlags();
}

function generateStatusResponse() {
  let response = {};
  for (let property in statusFlags) {
    response[property] = statusFlags[property] ? '1' : '0';
  }
  return Promise.wrap(response);
}

function init() {
  if (!statusFlags.prepared) {
    document.body.addEventListener('click', onPreviewClick);
    return VideoCapture.getCamerasAsync().then(function (cameras) {
      if (cameras.back && cameras.front) {
        statusFlags.canChangeCamera = true;
      }
      if (!cameras.back && !cameras.front) {
        return Promise.wrapError(errorTypes.CAMERA_UNAVAILABLE);
      }
      availableCameras = cameras;
      return initCamera().then(function () {
        statusFlags.prepared = true;
        statusFlags.authorized = true;
      });
    });
  }
  return Promise.wrap();
}

function initCamera(cameraType) {
  if (cameraType === cameraTypes.BACK && !availableCameras.back) {
    return Promise.wrapError(errorTypes.BACK_CAMERA_UNAVAILABLE);
  }
  if (cameraType === cameraTypes.FRONT && !availableCameras.front) {
    return Promise.wrapError(errorTypes.FRONT_CAMERA_UNAVAILABLE);
  }
  if (!cameraType) {
    if (availableCameras.front) {
      cameraType = cameraTypes.FRONT;
    }
    if (availableCameras.back) {
      cameraType = cameraTypes.BACK;
    }
  }
  preview.setVideoUrl('');
  return VideoCapture.createAsync(cameraType ? availableCameras.front : availableCameras.back).then(function (videoCapture) {
    currentVideoCapture = videoCapture;

    let videoUrl = URL.createObjectURL(currentVideoCapture.capture);

    if (statusFlags.showing) {
      preview.pause();
    }
    preview.setVideoUrl(videoUrl);
    if (statusFlags.showing) {
      preview.resume();
    }
    barcodeReader.setCapture(currentVideoCapture.capture);
    statusFlags.canEnableLight = currentVideoCapture.canEnableLight;
    statusFlags.currentCamera = cameraType;

  }, function (error) {
    const ACCESS_DENIED = -2147024891;
    if (error.number === ACCESS_DENIED) {
      statusFlags.denied = true;
      return Promise.wrapError(errorTypes.CAMERA_ACESS_DENIED);
    }
    return Promise.wrapError(errorTypes.UNEXPECTED_ERROR);
  });
}

function onPreviewClick(e) {
  if (statusFlags.showing && currentVideoCapture) {
    currentVideoCapture.focus();
  }
}

let qrScanner = {};

qrScanner.getStatus = function () {
  return init().then(generateStatusResponse, generateStatusResponse);
}

qrScanner.prepare = function () {
  return init().then(generateStatusResponse);
}

qrScanner.useCamera = function (inputStr) {
  return init().then(function () {
    let cameraType = parseInt(inputStr);
    return initCamera(cameraType).then(function () {
      return generateStatusResponse();
    });
  });
}

qrScanner.show = function () {
  return init().then(function () {
    preview.show();
    statusFlags.showing = true;
    statusFlags.previewing = preview.isPlaying();
    return generateStatusResponse();
  });
}

qrScanner.hide = function () {
  return init().then(function () {
    preview.hide();
    statusFlags.showing = false;
    statusFlags.previewing = false;
    return generateStatusResponse();
  });
}

let resolveLastScanPromise, rejectLastScanPromise;

qrScanner.scan = function () {

  if (statusFlags.scanning) {
    rejectLastScanPromise(errorTypes.SCAN_CANCELED);

    let lastScanPromise = new Promise(function (resolve, reject) {
      resolveLastScanPromise = resolve;
      rejectLastScanPromise = reject;
    });

    return lastScanPromise;

  }

  let lastScanPromise = new Promise(function (resolve, reject) {
    resolveLastScanPromise = resolve;
    rejectLastScanPromise = reject;
  });

  init().then(function () {
    barcodeReader.readCode().then(function (result) {
      if (!result) {
        return rejectLastScanPromise(errorTypes.SCAN_CANCELED);
      }
      resolveLastScanPromise(result.text);
      statusFlags.scanning = false;
    });
  }, function (error) {
    statusFlags.scanning = false;
    return rejectLastScanPromise(error);
  });

  statusFlags.scanning = true;

  return lastScanPromise;

}

qrScanner.cancelScan = function () {
  if (!statusFlags.scanning) return generateStatusResponse();
  statusFlags.scanning = false;
  barcodeReader.stop();
  return generateStatusResponse();
}

qrScanner.pausePreview = function () {
  preview.pause();
  statusFlags.previewing = false;
  return generateStatusResponse();
}

qrScanner.resumePreview = function () {
  preview.resume();
  statusFlags.previewing = statusFlags.showing;
  return generateStatusResponse();
}

//on Lumia devices, light functionality may be disabled while plugged in
qrScanner.enableLight = function () {
  return init().then(function () {
    if (statusFlags.lightEnabled) {
      return generateStatusResponse();
    }

    statusFlags.lightEnabled = currentVideoCapture.enableLight();

    if (!statusFlags.lightEnabled) {
      return Promise.wrapError(errorTypes.LIGHT_UNAVAILABLE);
    }

    return generateStatusResponse();
  });
}

qrScanner.disableLight = function () {

  if (statusFlags.lightEnabled) {
    currentVideoCapture.disableLight();
  }

  return generateStatusResponse();

}

qrScanner.openSettings = function () {
  return Promise.wrapError(errorTypes.OPEN_SETTINGS_UNAVAILABLE);
}

qrScanner.destroy = function () {
  reset();
  return generateStatusResponse();
}

reset();

function wrapPromise(fn) {
  return function (successCallback, errorCallback, strInput) {
    fn.call(window, strInput).then(successCallback, function (errorCode) {
      errorCallback(errorCode.toString() || '0');
    });
  }
}

for (let property in qrScanner) {
  if (typeof qrScanner[property] == "function") {
    exports[property] = wrapPromise(qrScanner[property])
  }
}

module.exports = exports;

cordova.commandProxy.add("QRScanner", exports);